### ISSUE TITLE
Enum class rejected by a generic union

### ISSUE THREAD

**Issue by kalekseev** [astral-sh/ty#4551]
### Summary

ty 0.0.82 rejects this call. ty 0.0.81 and Pyright 1.1.410 accept it.

```python
from collections.abc import Iterable
from enum import Enum

class Choice(Enum):
    A = "a"

def accepts_choices[T](choices: Iterable[tuple[T, str]] | type[Enum]) -> None:
    pass

accepts_choices(Choice)
```

```text
error[[invalid-argument-type](https://ty.dev/rules#invalid-argument-type)]: Argument to function `accepts_choices` is incorrect
  --> a.py:18:17
   |
18 | accepts_choices(Choice)
   |                 ^^^^^^ Expected `Iterable[tuple[Unknown, str]]`, found `<class 'Choice'>`
info: type `<class 'Choice'>` is not assignable to protocol `Iterable[tuple[Unknown, str]]`
info: └── protocol member `__iter__` is incompatible
info:     └── incompatible return types: `Iterator[Choice]` is not assignable to `Iterator[tuple[Unknown, str]]`
info:         └── protocol `Iterator[Choice]` is not assignable to protocol `Iterator[tuple[Unknown, str]]`
info:             └── protocol member `__next__` is incompatible
info:                 └── incompatible return types: `Choice` is not assignable to `tuple[Unknown, str]`
info: Function defined here
  --> a.py:14:5
   |
14 | def accepts_choices[T](choices: Iterable[tuple[T, str]] | type[Enum]) -> None:
   |     ^^^^^^^^^^^^^^^    --------------------------------------------- Parameter declared here
```

Expected: accept `Choice` through the `type[Enum]` union branch. That branch is absent from the diagnostic.

Replacing `T` with `str`, or using a legacy `TypeVar` with `default=str`, avoids the error.
This also affects Django field declarations with `choices=SomeTextChoices` or `choices=SomeIntegerChoices`.

### Version

ty 0.0.82 (6999edb88 2026-09-17)

**Comment by sharkdp** [comment]
Thank you for reporting this.

Seems like this regression was introduced in https://github.com/astral-sh/ruff/pull/28249 (FYI @dcreager)

Apparently this bug occurs because we (incorrectly) simplify `Iterable[tuple[T, str]] | type[Enum]` to `Iterable[tuple[T, str]]`. This is also why the error context tree doesn't show the second union element.

**Comment by carljm** [comment]
Further analysis suggests that the union remains intact in its original position in the signature of the generic function, but it gets reduced during call specialization -- we consider `T` unsolved (because `Choice` doesn't match the `Iterable` branch) and specialize it with `Unknown` -- which I think is right -- and at that point we consider `type[Enum]` to be redundant with `Iterable[tuple[Unknown, str]]` -- which is wrong -- so we simplify the specialized parameter type to just `Iterable[tuple[Unknown, str]]`.

So a simpler reproduction of the root cause is this:

```py
from collections.abc import Iterable
from enum import Enum
from typing import reveal_type, Any

def f(x: Iterable[tuple[Any, str]] | type[Enum]):
    reveal_type(x)  # Iterable[tuple[Any, str]]
```

But this incorrect redundancy judgment still comes from the constraint solver, because `EnumMeta.__iter__` is generic. Here's a Codex analysis of what happens:

<details>

The standalone annotation still invokes generic constraint solving because **`EnumMeta.__iter__` is generic**. The bug is an interaction between lazy redundancy checking and existential elimination:

1. [`EnumMeta.__iter__`](https://github.com/astral-sh/ruff/blob/b36c6cad203f2c2b1e7aea0b06fa49bfe9d8628e/crates/ty_vendored/vendor/typeshed/stdlib/enum.pyi#L142) has the signature `self: type[M] -> Iterator[M]`. Comparing `type[Enum]` with the iterable protocol produces:
   ```text
   Enum ≤ M
   M ≤ tuple[Unknown, str]
   ```
   The first constraint comes from binding the receiver; the second comes from comparing iterator element types.

2. The [lazy type-variable comparison](https://github.com/astral-sh/ruff/blob/b36c6cad203f2c2b1e7aea0b06fa49bfe9d8628e/crates/ty_python_semantic/src/types/relation.rs#L1717-L1728) preserves `tuple[Unknown, str]` unchanged for **redundancy**. It only materializes the bound when the relation is **subtyping**.

3. [`add_sequents_with_concrete_upper`](https://github.com/astral-sh/ruff/blob/b36c6cad203f2c2b1e7aea0b06fa49bfe9d8628e/crates/ty_python_semantic/src/types/constraints/sequents.rs#L1342-L1348) derives consequences from these two bounds only when both are fully static. The nested `Unknown` causes it to skip deriving their contradiction.

4. [`reduce_inferable`](https://github.com/astral-sh/ruff/blob/b36c6cad203f2c2b1e7aea0b06fa49bfe9d8628e/crates/ty_python_semantic/src/types/constraints.rs#L735-L768) existentially eliminates `M`. Its Boolean abstraction removes the constraint nodes, relying on derived facts to preserve their compatibility requirements. Because the contradiction was never derived, it incorrectly returns **true**, and the union builder removes `type[Enum]`.

I verified the solver failure directly:

```text
∃M. Enum ≤ M ∧ M ≤ tuple[Unknown, str]  → True   # wrong
∃M. Enum ≤ M ∧ M ≤ tuple[str, str]      → False
```

This also explains why the standalone **subtyping** check correctly returns false: it first materializes the upper bound to `tuple[Never, str]`, allowing the contradiction to be detected.

The underlying defect is that **incomplete constraint elimination is being used as proof of redundancy**. The static-bound guard prevents unsafe gradual transitivity, but skipping that reasoning must not make an impossible existential constraint count as proven true.

</details>

There seems to be some overlap here with the direction of https://github.com/astral-sh/ruff/pull/28657 (though that alone doesn't fix this) and the ongoing scoped-quantification work, but I'm not sure if this is "will definitely be fixed by that work" or "new problem needs dedicated attention"; cc @ibraheemdev @dcreager.

**Comment by carljm** [comment]
Confirmed offline with @dcreager that the best course of action here is to wait for existing pending work to land, and then see what the remaining gap is, because both steps 3 and 4 in the Codex analysis above correspond directly to ongoing work.

**Comment by charliermarsh** [comment]
I'm going to assign to @dcreager so this doesn't look "available".
